### ISSUE TITLE
no-dupe-keys breaks with experimentalObjectRestSpread & AssignmentExpression

### ISSUE THREAD

**Issue by gre** [eslint/eslint#3042]
After the https://github.com/eslint/espree/issues/167 fix,

I still can't make the experimentalObjectRestSpread working with AssignmentExpression when using with `no-dupe-keys` rule. **if I disable no-dupe-keys rule, the error disappear**, so I think this is related to this rule.

Here is what happens to me:

```
$ npm install -g eslint@1.0.0-rc-1     # this grabs espree 2.2.2
...
$ eslint test.js
/Users/gre/.nvm/versions/io.js/v2.3.4/lib/node_modules/eslint/lib/rules/no-dupe-keys.js:25
                var keyName = property.key.name || property.key.value,
                                          ^
TypeError: Cannot read property 'name' of undefined
    at /Users/gre/.nvm/versions/io.js/v2.3.4/lib/node_modules/eslint/lib/rules/no-dupe-keys.js:25:43
    at Array.forEach (native)
    at EventEmitter.ObjectExpression (/Users/gre/.nvm/versions/io.js/v2.3.4/lib/node_modules/eslint/lib/rules/no-dupe-keys.js:24:29)
    at emitOne (events.js:82:20)
    at EventEmitter.emit (events.js:169:7)
    at Controller.controller.traverse.enter (/Users/gre/.nvm/versions/io.js/v2.3.4/lib/node_modules/eslint/lib/eslint.js:790:25)
    at Controller.__execute (/Users/gre/.nvm/versions/io.js/v2.3.4/lib/node_modules/eslint/node_modules/estraverse/estraverse.js:397:31)
    at Controller.traverse (/Users/gre/.nvm/versions/io.js/v2.3.4/lib/node_modules/eslint/node_modules/estraverse/estraverse.js:495:28)
    at EventEmitter.module.exports.api.verify (/Users/gre/.nvm/versions/io.js/v2.3.4/lib/node_modules/eslint/lib/eslint.js:783:24)
    at processText (/Users/gre/.nvm/versions/io.js/v2.3.4/lib/node_modules/eslint/lib/cli-engine.js:199:27)
```

test.js

```
const a = { foo: {} };

const b = {
  ...a.foo,
  a
};
```

.eslintrc

```
{
    "rules": {
        "indent": [
            2,
            4
        ],
        "quotes": [
            2,
            "double"
        ],
        "linebreak-style": [
            2,
            "unix"
        ],
        "semi": [
            2,
            "always"
        ]
    },
    "env": {
        "es6": true,
        "browser": true
    },
    "extends": "eslint:recommended",
    "ecmaFeatures": {
      "jsx": true,
      "experimentalObjectRestSpread": true
    }
}
```

Thanks

**Comment by nzakas** [comment]
Definitely a bug. Working on this.

**Comment by nzakas** [comment]
Fixed in #3051
